lib/psabpf_pipeline.c

```diff
@@ -113,175 +160,179 @@ static int xdp_port_add(__u32 pipeline_id, char *intf)
+    /* Do not close fd obtained from above call, it is maintained by obj */
     if (ret < 0 || obj == NULL) {
-        fprintf(stderr, "cannot load the BPF program, code = %d\n", ret);
-        return -1;
+        ret = errno;
+        fprintf(stderr, "cannot load the BPF program: %s\n", strerror(ret));
+        return ret;
     }
 
     bpf_object__for_each_program(pos, obj) {
         const char *sec_name = bpf_program__section_name(pos);
-        int prog_fd = bpf_program__fd(pos);
+        fd = bpf_program__fd(pos);
         if (!strcmp(sec_name, TC_INIT_PROG) || !strcmp(sec_name, XDP_INIT_PROG)) {
-            ret = do_initialize_maps(prog_fd);
+            ret = do_initialize_maps(fd);
             if (ret) {
+                ret = -errno;
+                fprintf(stderr, "failed to initialize maps: %s\n", strerror(errno));
                 goto err_close_obj;
             }
             // do not pin map initializer
             continue;
         }
 
-        snprintf(pinned_file, sizeof(pinned_file), "%s/%s%d/%s", BPF_FS,
-                 PIPELINE_PREFIX, pipeline->id, __bpf_program__pin_name(pos));
+        build_ebpf_prog_filename(pinned_file, sizeof(pinned_file),
+                                 ctx, program_pin_name(pos));
 
         ret = bpf_program__pin(pos, pinned_file);
         if (ret < 0) {
+            fprintf(stderr, "failed to pin %s at %s: %s\n",
+                    sec_name, pinned_file, strerror(-ret));
             goto err_close_obj;
         }
     }
 
     struct bpf_map *map;
     bpf_object__for_each_map(map, obj) {
         if (bpf_map__is_pinned(map)) {
-            if (bpf_map__unpin(map, NULL)) {
+            ret = bpf_map__unpin(map, NULL);
+            if (ret) {
+                fprintf(stderr, "failed to remove old map pin file: %s\n", strerror(-ret));
                 goto err_close_obj;
             }
         }
 
-        memset(pinned_file, 0, sizeof(pinned_file));
-        snprintf(pinned_file, sizeof(pinned_file), "%s/%s%d/%s/%s", BPF_FS,
-                 PIPELINE_PREFIX, pipeline->id, "maps", bpf_map__name(map));
-        if (bpf_map__set_pin_path(map, pinned_file)) {
+        const char *map_name = bpf_map__name(map);
+
+        /* Pinned file name cannot contain a dot */
+        if (strstr(map_name, ".") != NULL)
+            continue;
+
+        build_ebpf_map_filename(pinned_file, sizeof(pinned_file), ctx, map_name);
+        ret = bpf_map__set_pin_path(map, pinned_file);
+        if (ret) {
+            fprintf(stderr, "failed to pin map at %s: %s\n", pinned_file, strerror(-ret));
             goto err_close_obj;
         }
 
-        if (bpf_map__pin(map, pinned_file)) {
+        ret = bpf_map__pin(map, pinned_file);
+        if (ret) {
+            fprintf(stderr, "failed to pin map at %s: %s\n", pinned_file, strerror(-ret));
             goto err_close_obj;
         }
     }
 
 err_close_obj:
     bpf_object__close(obj);
 
-    return ret;
+    return -ret;
```

Comment:
Why minus?

Reply:
Generally speaking:
- sometimes `libbpf` return error codes in `errno`, they are positive values (especially wrappers for system call);
- sometimes `libbpf` return error codes as a result of function, they are negative values (this case) - see [libbpf source code](https://github.com/libbpf/libbpf/blob/v0.3/src/libbpf.c#L7758-L7766) for reference.

If we assume that our library returns positive error codes, we must make error codes positive from negative ones.